mayastor/tests/aio_unmap.rs

```diff
@@ -0,0 +1,118 @@
+use common::MayastorTest;
+use mayastor::{
+    core::{Bdev, MayastorCliArgs},
+    lvs::Lvs,
+};
+use rpc::mayastor::CreatePoolRequest;
+use std::{ffi::CString, io::Error, mem::MaybeUninit};
+
+pub mod common;
+
+static DISKNAME1: &str = "/tmp/disk1.img";
+
+fn get_allocated_blocks(path: &str) -> Result<i64, Error> {
+    let mut data: MaybeUninit<libc::stat64> = MaybeUninit::uninit();
+    let cpath = CString::new(path).unwrap();
+
+    if unsafe { libc::stat64(cpath.as_ptr(), data.as_mut_ptr()) } < 0 {
+        return Err(Error::last_os_error());
+    }
+
+    Ok(unsafe { data.assume_init().st_blocks })
+}
+
+#[tokio::test] #[ignore] // needs updated spdk
+async fn aio_unmap_test() {
+    common::delete_file(&[DISKNAME1.into()]);
+    common::truncate_file(DISKNAME1, 64 * 1024);
+
+    // Verify that there are currently no blocks allocated
+    // for the sparse file that is our backing store.
+    assert_eq!(get_allocated_blocks(DISKNAME1).unwrap(), 0);
+
+    let args = MayastorCliArgs {
+        reactor_mask: "0x3".into(),
+        ..Default::default()
+    };
+    let ms = MayastorTest::new(args);
+
+    // Create a pool.
+    ms.spawn(async {
+        Lvs::create_or_import(CreatePoolRequest {
+            name: "tpool".into(),
+            disks: vec!["aio:///tmp/disk1.img".into()],
+        })
+        .await
+        .unwrap();
+    })
+    .await;
+
+    // Check that we're able to find our new LVS.
+    ms.spawn(async {
+        assert_eq!(Lvs::iter().count(), 1);
+        let pool = Lvs::lookup("tpool").unwrap();
+        assert_eq!(pool.name(), "tpool");
+        assert_eq!(pool.used(), 0);
+        dbg!(pool.uuid());
+        assert_eq!(pool.base_bdev().name(), "/tmp/disk1.img");
+    })
+    .await;
+
+    // Create 4 lvols on this pool.
+    ms.spawn(async {
+        let pool = Lvs::lookup("tpool").unwrap();
+        for i in 0 .. 4 {
+            pool.create_lvol(&format!("vol-{}", i), 16 * 1024, true)
+                .await
+                .unwrap();
+        }
+
+        let pool = Lvs::lookup("tpool").unwrap();
+        assert_eq!(pool.lvols().unwrap().count(), 4);
+    })
+    .await;
+
+    // verify that some blocks have been allocated
+    assert_ne!(get_allocated_blocks(DISKNAME1).unwrap(), 0);
```

Comment:
Looks like these lvols only have metadata, should we also write some "user data" to the lvol?
Would it be worth sending explicit unmap to each lvol or even via a nexus? (to cover the case for unmaps from the pod host)?

Reply:
Yes we could write user data to the lvols. However this was not supposed to be a test to validate that the upper layers do the right thing when the lvol or pool is destroyed. This was intended to be a simple test to validate that when the upper layers want to zero a region, the blocks comprising that region are in fact unallocated. The fact that everything is zeroed means I don't need to complicate things by having to verify precisely where a hole is located in a file.

Reply:
After discussing with tiago, will create 2 tests:
- low level tests that verifies calling unmap on a bdev actually unallocates the blocks form the underlying file.
- a higher level test that verifies that destroying an lvol or pool does in fact zero the data (without being concerned as to how the zeroes are written).